modin/pandas/groupby.py

```diff
@@ -270,7 +270,10 @@ def apply(self, func, *args, **kwargs):
     def dtypes(self):
         if self._axis == 1:
             raise ValueError("Cannot call dtypes on groupby with axis=1")
-        return self._apply_agg_function(lambda df: df.dtypes, drop=self._as_index)
+        if self._is_multi_by and not self._as_index:
```

Comment:
No, it's important to call `apply` function, because otherwise we can't retrieve `dtypes` from 'by' columns:

```python
# When we do:
self._apply_agg_function(lambda df: df.dtypes)
# That means that inside `groupby_agg_builder` we will apply 
# passed lambda to our pandas groupby object:
df.groupby(by, **groupby_kwargs).dtypes
```
And because we're always set `as_index=True` to track metadata, we will always get mismatched 'by' columns, even if the original `as_index` value was `False`:
```python
>>> pd_df
  col0  col1  col2 col3
0    a     1     5  one
1    b     2     6  one
2    a     3     7  two
3    a     4     8  two
>>> pd_df.groupby(["col0", "col3"], as_index=False).dtypes  # That's the result we want
     col0   col1   col2    col3
0  object  int64  int64  object
1  object  int64  int64  object
2  object  int64  int64  object
>>> pd_df.groupby(["col0", "col3"], as_index=True).dtypes  # That's the result we'll get.
            col1   col2					   # As we can see, there is no way
col0 col3                                                  # to restore 'col0' and 'col3' dtypes values
a    one   int64  int64
     two   int64  int64
b    one   int64  int64
```
So the solution here is to use `apply` function, that does not drop 'by' columns from frame:
```python
>>> pd_df.groupby(["col0", "col3"], as_index=True).apply(lambda df: df.dtypes)
             col0   col1   col2    col3
col0 col3
a    one   object  int64  int64  object
     two   object  int64  int64  object
b    one   object  int64  int64  object

# So what we want to do here is:
self._apply_agg_function(lambda grp: grp.apply(lambda df: df.dtypes))
# Which is the same as:
self.apply(lambda df: df.dtypes)
```
Note: `drop` parameter, that indicate whether we should drop 'by' column before aggregation or not, works only in case of single-column groupby: https://github.com/modin-project/modin/blob/e5556b5c531b52546a76690bc976aa90995862a7/modin/pandas/groupby.py#L879-L885

Reply:
Got it.